# Inventario inicial de los datos


## Análisis de los datos de 2023

In [ ]:
# Importar librerías principales
import pandas as pd

In [2]:
# Listado de rutas de archivos Excel en la carpeta raw
from pathlib import Path

RAW_PATH = Path("../data/raw")

excel_files = list(RAW_PATH.rglob("*.xlsx")) + list(RAW_PATH.rglob("*.xls"))

len(excel_files), excel_files

(3,
 [WindowsPath('../data/raw/altura_min_sevilla.xlsx'),
  WindowsPath('../data/raw/datos_2023.xlsx'),
  WindowsPath('../data/raw/datos_2024.xlsx')])

In [3]:
# Creación de un inventario de los archivos Excel encontrados
inventory = []

for file in excel_files:
    try:
        df = pd.read_excel(file)
        
        inventory.append({
            "file_name": file.name,
            "file_path": str(file),
            "folder": file.parent.name,
            "n_rows": df.shape[0],
            "n_columns": df.shape[1],
            "columns": list(df.columns),
            "n_missing_total": df.isna().sum().sum(),
            "missing_percentage": df.isna().sum().sum() / (df.shape[0] * df.shape[1]) * 100
        })
        
    except Exception as e:
        inventory.append({
            "file_name": file.name,
            "file_path": str(file),
            "folder": file.parent.name,
            "n_rows": None,
            "n_columns": None,
            "columns": None,
            "n_missing_total": None,
            "missing_percentage": None,
            "error": str(e)
        })

inventory_df = pd.DataFrame(inventory)
inventory_df.head()

,file_name,file_path,folder,n_rows,n_columns,columns,n_missing_total,missing_percentage
0,altura_min_sevilla.xlsx,..\data\raw\altura_min_sevilla.xlsx,raw,525601,7,"[Fecha, Año, Mes, Dia, Hora, Min, altura]",2,0.000054
1,datos_2023.xlsx,..\data\raw\datos_2023.xlsx,raw,525600,16,"[Ano, Mes, Dia, Hora, Minuto, Fecha, GHI, DNI,...",87769,1.043676
2,datos_2024.xlsx,..\data\raw\datos_2024.xlsx,raw,527040,15,"[Mes, Dia, Hora, Minuto, Fecha, GHI, DNI, DHI,...",328539,4.155776


In [4]:
inventory_df.to_csv("../reports/tables/data_inventory.csv", index=False)

In [5]:
sample_2023 = [f for f in excel_files if "2023.xlsx" in str(f)][0]
sample_2024 = [f for f in excel_files if "2024.xlsx" in str(f)][0]

df_2023 = pd.read_excel(sample_2023)
df_2024 = pd.read_excel(sample_2024)

display(df_2023.head())

,Ano,Mes,Dia,Hora,Minuto,Fecha,GHI,DNI,DHI,Temperatura [°C],Velocidad del viento [m/s],Direccion del viento [º],Humedad Relativa [%],Codigo GHI,Codigo DNI,Codigo dHI
0,2023,1,1,0,0,2023-01-01 00:00:00,0.0,0.0,0.0,10.0,1.050,117.750,77.0,0,0,0
1,2023,1,1,0,1,2023-01-01 00:01:00,0.0,0.0,0.0,10.0,1.333,311.500,77.0,0,0,0
2,2023,1,1,0,2,2023-01-01 00:02:00,0.0,0.0,0.0,10.0,1.100,256.417,77.0,0,0,0
3,2023,1,1,0,3,2023-01-01 00:03:00,0.0,0.0,0.0,10.0,1.333,290.417,77.0,0,0,0
4,2023,1,1,0,4,2023-01-01 00:04:00,0.0,0.0,0.0,10.0,1.058,333.667,77.0,0,0,0


## Variables del dataset
- **Año**: Año en el que se registró la observación.
- **Mes**: Mes del año en el que se tomó la medida.
- **Día**: Día del mes correspondiente al registro.
- **Hora**: Hora del día en la que se realizó la medición.
- **Minuto**: Minuto exacto en el que se realizó la medición.
- **Fecha**: Marca temporal completa correspondiente a la observación.
- **GHI**: Irradiancia solar global medida sobre una superficie horizontal.
- **DNI**: Irradiancia solar directa medida sobre una superficie perpendicular a los rayos solares.
- **DHI**: Irradiancia solar difusa recibida sobre una superficie horizontal.
- **Temperatura [°C]**: Temperatura ambiente registrada, expresada en grados Celsius.
- **Velocidad del viento [m/s]**: Velocidad del viento registrada, expresada en metros por segundo.
- **Dirección del viento [°]**: Dirección de procedencia del viento, expresada en grados.
- **Humedad relativa [%]**: Humedad relativa del aire en el momento de la medición, expresada como porcentaje.
- **Código GHI**: Variable objetivo con valores entre 0 y 1 que indica la calidad o validez de la medición de GHI.
- **Código DNI**: Variable objetivo con valores entre 0 y 2 que indica la calidad o validez de la medición de DNI.
- **Código DHI**: Variable objetivo con valores entre 0 y 2 que indica la calidad o validez de la medición de DHI.

In [6]:
display(df_2024.head())

,Mes,Dia,Hora,Minuto,Fecha,GHI,DNI,DHI,Temperatura [°C],Velocidad del viento [m/s],Direccion del viento [º],Humedad Relativa [%],Codigo GHI,Codigo DNI,Codigo dHI
0,1,1,0,0,2024-01-01 00:00:00,0.0,0.0,0.0,9.2,0.925,312.167,92.000,0,0,0
1,1,1,0,1,2024-01-01 00:01:00,0.0,0.0,0.0,9.2,1.125,310.833,92.167,0,0,0
2,1,1,0,2,2024-01-01 00:02:00,0.0,0.0,0.0,9.2,1.233,304.333,92.417,0,0,0
3,1,1,0,3,2024-01-01 00:03:00,0.0,0.0,0.0,9.2,1.083,313.500,92.000,0,0,0
4,1,1,0,4,2024-01-01 00:04:00,0.0,0.0,0.0,9.2,1.050,311.000,92.333,0,0,0


El año 2024 no incluye la columna de año como 2023 pero el resto de columnas son las mismas

In [7]:
df_2023.info()

<class 'pandas.DataFrame'>
RangeIndex: 525600 entries, 0 to 525599
Data columns (total 16 columns):
 #   Column                      Non-Null Count   Dtype         
---  ------                      --------------   -----         
 0   Ano                         525600 non-null  int64         
 1   Mes                         525600 non-null  int64         
 2   Dia                         525600 non-null  int64         
 3   Hora                        525600 non-null  int64         
 4   Minuto                      525600 non-null  int64         
 5   Fecha                       525600 non-null  datetime64[us]
 6   GHI                         523813 non-null  float64       
 7   DNI                         523813 non-null  float64       
 8   DHI                         523813 non-null  float64       
 9   Temperatura [°C]            504998 non-null  float64       
 10  Velocidad del viento [m/s]  504998 non-null  float64       
 11  Direccion del viento [º]    504998 non-null  float

Se encuentran nulos en todas las features no relacionadas con la fecha, incluyendo la toma de los valores de las 3 irradaciones. 

Existen 525600 muestras de datos correspondientes a una medida por minuto de los 365 días de 2023.

Todas las variables excepto la fecha completa son numéricas.

In [8]:
df_2024.info()

<class 'pandas.DataFrame'>
RangeIndex: 527040 entries, 0 to 527039
Data columns (total 15 columns):
 #   Column                      Non-Null Count   Dtype         
---  ------                      --------------   -----         
 0   Mes                         527040 non-null  int64         
 1   Dia                         527040 non-null  int64         
 2   Hora                        527040 non-null  int64         
 3   Minuto                      527040 non-null  int64         
 4   Fecha                       527040 non-null  datetime64[us]
 5   GHI                         516959 non-null  float64       
 6   DNI                         516959 non-null  float64       
 7   DHI                         516959 non-null  float64       
 8   Temperatura [°C]            452466 non-null  float64       
 9   Velocidad del viento [m/s]  452466 non-null  float64       
 10  Direccion del viento [º]    452466 non-null  float64       
 11  Humedad Relativa [%]        452466 non-null  float

Las features con nulos son las mismas que para el caso de 2023 pero existe una gran mayor parte de nulos en las condiciones meteorológicas de 2024 que para 2023 además de más nulos en las mediciones de irradación. 

En este caso se tienen 527040 muestras de datos debido a que 2024 fue año bisiesto por lo que se tienen 366 días en el año.

Todas las variables excepto la fecha completa son numéricas.

In [14]:
date_cols = ["Ano", "Mes", "Dia", "Hora", "Minuto", "Fecha"]
target_code_cols = ["Codigo GHI", "Codigo DNI", "Codigo dHI"]

numeric_feature_cols_2023 = [
    col for col in df_2023.columns 
    if col not in date_cols + target_code_cols
]

df_2023[numeric_feature_cols_2023].describe().T

,count,mean,std,min,25%,50%,75%,max
GHI,523813.0,171.425955,278.596937,0.000,0.000,0.000,279.917,2525.500
DNI,523813.0,247.646448,350.730151,0.000,0.000,0.000,600.333,1083.250
DHI,523813.0,77.933365,145.115104,0.000,0.000,0.000,90.083,1072.500
Temperatura [°C],504998.0,20.237443,8.298371,0.517,14.150,19.825,25.900,43.450
Velocidad del viento [m/s],504998.0,1.924322,1.354026,0.000,0.975,1.625,2.575,19.442
Direccion del viento [º],504998.0,177.157477,97.548361,0.000,74.000,197.583,247.917,355.000
Humedad Relativa [%],504998.0,54.518707,22.856489,6.000,35.000,55.667,74.000,94.667


Las features de condiciones meteorológicas no presentan valores negativos, al igual que las irradancias, por lo que en ese aspecto no se han tomado valores incorrectos (teniendo en cuenta que la temperatura sí podría ser negativa aunque en Sevilla es poco factible)

In [15]:
numeric_feature_cols_2024 = [
    col for col in df_2024.columns 
    if col not in date_cols + target_code_cols
]

df_2024[numeric_feature_cols_2024].describe().T

,count,mean,std,min,25%,50%,75%,max
GHI,516959.0,185.210492,292.316467,0.0,0.000,0.000,307.583,1448.000
DNI,516959.0,209.881400,330.913676,0.0,0.000,0.000,453.364,1000.083
DHI,516959.0,62.269301,124.748119,0.0,0.000,0.000,72.000,1153.000
Temperatura [°C],452466.0,19.633798,7.417307,-50.0,14.300,18.992,24.300,42.400
Velocidad del viento [m/s],452466.0,1.892488,1.478352,0.0,0.878,1.630,2.642,14.725
Direccion del viento [º],452466.0,179.661096,97.794484,0.0,74.583,200.583,249.583,355.000
Humedad Relativa [%],452466.0,60.417413,20.700741,0.0,45.000,63.000,78.000,95.000


En este caso si se presenta una temperatura mínima de -50ºC que habrá que analizar ya que no es factible en Sevilla, pero el resto de variables no parecen presentar problemas de mediciones 

In [16]:
# Porcentaje de nulos por columna en archivo ejemplo
missing_2023 = df_2023[numeric_feature_cols_2023].isna().mean().sort_values(ascending=False) * 100
missing_2023

Direccion del viento [º]      3.919711
Velocidad del viento [m/s]    3.919711
Temperatura [°C]              3.919711
Humedad Relativa [%]          3.919711
DHI                           0.339992
GHI                           0.339992
DNI                           0.339992
dtype: float64

In [17]:
missing_2024 = df_2024[numeric_feature_cols_2024].isna().mean().sort_values(ascending=False) * 100
missing_2024

Direccion del viento [º]      14.149590
Velocidad del viento [m/s]    14.149590
Temperatura [°C]              14.149590
Humedad Relativa [%]          14.149590
DHI                            1.912758
GHI                            1.912758
DNI                            1.912758
dtype: float64

In [26]:
print("Duplicados fecha 2023:", df_2023["Fecha"].duplicated().sum())
print("Duplicados fecha 2024:", df_2024["Fecha"].duplicated().sum())

print("2023 ordenado:", df_2023["Fecha"].is_monotonic_increasing)
print("2024 ordenado:", df_2024["Fecha"].is_monotonic_increasing)

Duplicados fecha 2023: 0
Duplicados fecha 2024: 0
2023 ordenado: True
2024 ordenado: True


No existen fechas duplicadas por lo que tenemos justo una medida por minuto durante todos los días del año para ambos años. Además, estos datos están perfectamente ordenados. 

De esta manera, no es necesario actuar sobre las medidas diarias a nivel de falta de medidas para algún minuto en específico, recordando que algunas medidas no tienen datos meteorológicos

In [27]:
for col in ["Codigo GHI", "Codigo DNI", "Codigo dHI"]:
    print(col)
    print("2023")
    print(df_2023[col].value_counts(dropna=False).sort_index())
    print("2024")
    print(df_2024[col].value_counts(dropna=False).sort_index())
    print()

Codigo GHI
2023
Codigo GHI
0    462926
1     62674
Name: count, dtype: int64
2024
Codigo GHI
0    437044
1     89996
Name: count, dtype: int64

Codigo DNI
2023
Codigo DNI
0    503758
1     20760
2      1082
Name: count, dtype: int64
2024
Codigo DNI
0    439910
1     86287
2       843
Name: count, dtype: int64

Codigo dHI
2023
Codigo dHI
0    467954
1     57464
2       182
Name: count, dtype: int64
2024
Codigo dHI
0    392052
1    134782
2       206
Name: count, dtype: int64



In [13]:
temp_neg = df_2024[df_2024["Temperatura [°C]"] < 0][['Fecha', 'Temperatura [°C]']].sort_values(by='Temperatura [°C]', ascending=True)
temp_neg

,Fecha,Temperatura [°C]
111959,2024-03-18 17:59:00,-50.000
113459,2024-03-19 18:59:00,-50.000
113464,2024-03-19 19:04:00,-50.000
113467,2024-03-19 19:07:00,-50.000
114730,2024-03-20 16:10:00,-50.000
114738,2024-03-20 16:18:00,-50.000
117333,2024-03-22 11:33:00,-50.000
117335,2024-03-22 11:35:00,-50.000
117336,2024-03-22 11:36:00,-50.000
117356,2024-03-22 11:56:00,-50.000


In [ ]:
alturas = pd.read_excel("../data/raw/alturas_sevilla.xlsx")
alturas.head()

,Fecha,Año,Mes,Dia,Hora,Min,altura
0,2023-01-01 00:00:00,2023,1.0,1,0,0,0.0
1,2023-01-01 00:01:00,2023,1.0,1,0,1,0.0
2,2023-01-01 00:02:00,2023,1.0,1,0,2,0.0
3,2023-01-01 00:03:00,2023,1.0,1,0,3,0.0
4,2023-01-01 00:04:00,2023,1.0,1,0,4,0.0


In [21]:
alturas.info()

<class 'pandas.DataFrame'>
RangeIndex: 525601 entries, 0 to 525600
Data columns (total 7 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   Fecha   525601 non-null  datetime64[us]
 1   Año     525601 non-null  int64         
 2   Mes     525600 non-null  float64       
 3   Dia     525601 non-null  int64         
 4   Hora    525601 non-null  int64         
 5   Min     525601 non-null  int64         
 6   altura  525600 non-null  float64       
dtypes: datetime64[us](1), float64(2), int64(4)
memory usage: 28.1 MB


Existe dos valores faltantes, uno para la columna Mes y otro para altura.

In [35]:
alturas[alturas.isna().any(axis=1)]

,Fecha,Año,Mes,Dia,Hora,Min,altura
525600,2024-01-01,2023,NaN,365,24,0,NaN


In [36]:
alturas.tail()

,Fecha,Año,Mes,Dia,Hora,Min,altura
525596,2023-12-31 23:56:00,2023,12.0,365,23,56,0.0
525597,2023-12-31 23:57:00,2023,12.0,365,23,57,0.0
525598,2023-12-31 23:58:00,2023,12.0,365,23,58,0.0
525599,2023-12-31 23:59:00,2023,12.0,365,23,59,0.0
525600,2024-01-01 00:00:00,2023,NaN,365,24,0,NaN


Se trata de una medida correspondiente al comienzo del año 2024, por lo que es un parámetro que se puede eliminar si se quiere tener la información únicamente de 2023

In [ ]:
alturas['altura'].describe()

count    525600.000000
mean         16.121101
std          21.212558
min           0.000000
25%           0.000000
50%           0.220741
75%          29.994233
max          76.031723
Name: altura, dtype: float64